### ISSUE TITLE
[DX] Allow modules to declare (some of) the hooks they implement in their .info files, so that they can be discovered by module_implements() even when they are disabled

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#6478]
This is related to #5367 and #6045.

In [this Zulip thread](https://backdrop.zulipchat.com/#narrow/stream/218635-Backdrop/topic/discover.20hook.20implementations.20provided.20by.20disabled.20modules) I concluded that the problem I was trying to solve (discover hook implementations in _disabled_ modules) could be addressed by allowing modules to declare (even some of) the hooks they implement in their .info files:

@klonos 
> I'm already aware of `module_implements()`, however that function discovers hook implementations provided by _enabled_ modules. What I need is a way to discover hook implementations provided by modules that have not been enabled yet (but exist in the codebase).
>
> ...for example, the second set of settings in `admin/config/development/logging` (the severity levels) only make sense if either dblog or syslog are enabled - there is no point otherwise. Now this is with regards to core-provided modules, but there may be contrib modules out there implementing `hook_watchdog()` ...
>
> I was thinking that we should be hiding these settings, and add a message along the following lines:
>> There are currently no enabled modules that provide any logging capabilities. Consider enabling any of the following modules to see relevant settings here: ...
>
> However it is not possible to get a list of modules implementing `hook_watchdog()` if these modules are disabled. If you enable them, well then you show these settings, so there's no point now.

@argiepiano 
> ... What you describe would be extremely expensive when someone has a lot of "disabled" modules in the folder. ...

@klonos 
> Thanks @argiepiano ...I have looked at the code and considered how PHP and Drupal works, and I have also reached to the same conclusion. Just thought that I'd ask, in case there's something that I might be missing.

...

> ...given this some final thought before giving up on the idea, and the only thing that I could come up with that could work is if we had a way for modules to declare (even some of) the hooks they implement via their .info files. Since .info data is being parsed even for disabled modules, the list of hooks would then be available with things like `system_rebuild_module_data()`.

**Comment by klonos** [comment]
Rehashing some feedback from @quicksketch left in the Zulip thread:
> To answer the main question: I don't think this is easily possible. You could read in the .module files as text and parse them (that's effectively what Doctrine used to provide, that Drupal 8+ used, see https://github.com/Roave/BetterReflection for a current example). But that wouldn't be 100% accurate since modules can have hooks outside of the .module file by using hook_hook_info().
>
>The basic premise here seems like a bit of an edge-case though. You can already identify if no active modules implement a hook, so you could display a message or status report item already. So the only part you're missing is suggesting other modules. In that situation, you could either link to documentation or provide the most common modules as suggestions. That would especially work in situations where a single project bundled multiple sub-modules, one of which probably has the recommended hook.

**Comment by argiepiano** [comment]
> There are other files that implement hooks in contrib modules, e.g.,

Plus all the `install` files.

**Comment by klonos** [comment]
Yup, that's why I propose that we use .info files as the means to build some sort of (partial and selective) registry of specific hooks implemented by modules, which would allow these specific hooks to be detected as "conditionally" available (only when the respective module is enabled).

That would be something completely optional, so it can only be implemented in an opt-in manner (not all modules need to declare any hook in their .info, and if they do, then they don't have to declare every single hook they implement - only declare specific hooks in the .info file, and only if there is a use case for it). I'm thinking that this would be somewhat a similar approach to what we do with .test.info files.

Since data from .info files is already collected, saved, and readily available, we won't have to do anything extra (so it will not be an "expensive" task). Functions like `module_implements()` could then get an extra `$include_disabled` parameter (API addition - not a breaking change), which will default to `FALSE` or a `$scope` parameter, which would default to `'enabled'` but also allow `'disabled'` and `'both'` as alternative values. If we don't want to mess with existing functions, we could then create some new `module_implements_disabled()` function or something like that.

Anyway, point is that this sounds like something interesting to work on, and I'll loop back to this when I want something to experiment with.

PS: the other day, I thought of another use case in core where this would be useful, but now forget what it was. If I remember, I'll come back to add it here.

**Comment by klonos** [comment]
...I remembered the use case: `filter_get_editors()` only retrieves the text editors that are available with enabled modules (it scans for implementations of `hook_editor_info()`). So if you wanted to say add a message in the status report page or above the `admin/config/content/formats` as help text (to improve the discoverability of editors provided by disabled modules), then there is currently no way to do that.

Also, having the hook registered in the .info file of modules might allow or make it easier to be able to filter the list of modules in the project browser by modules that provide alternative text editors. ...one could argue that this can be achieved with tags/categories in the .info files though.
